queue/dependencytracker.js

```diff
@@ -0,0 +1,279 @@
+ * tasks are scheduled.
+ *
+ *
+ * Options:
+ * {
+ *   Task:              data.Task instance
+ *   publisher:         publisher from exchanges
+ *   queueService:      QueueService instance
+ *   TaskDependency:    data.TaskDependency instance
+ *   TaskRequirement:   data.TaskRequirement instance
+ * }
+ */
+class DependencyTracker {
+  constructor(options = {}) {
+    // Validate options
+    assert(options.Task,            'Expected options.Task');
+    assert(options.publisher,       'Expected options.publisher');
+    assert(options.queueService,    'Expected options.queueService');
+    assert(options.TaskDependency,  'Expected options.TaskDependency');
+    assert(options.TaskRequirement, 'Expected options.TaskRequirement');
+
+    // Store options on this object
+    this.Task             = options.Task;
+    this.publisher        = options.publisher;
+    this.queueService     = options.queueService;
+    this.TaskDependency   = options.TaskDependency;
+    this.TaskRequirement  = options.TaskRequirement;
+  }
+
+  /**
+   * Track dependencies for task object, and add pending run to the task object,
+   * if all dependencies was resolved during creation.
+   *
+   * This will return {message, details} if there is an error.
+   */
+  async trackDependencies(task) {
+
+    // Create TaskRequirement entries, each entry implies that taskId is blocked
+    // by requiredTaskId. This relation is used to track if a taskId is blocked.
+    await Promise.all(task.dependencies.map(requiredTaskId => {
+      return this.TaskRequirement.create({
+        taskId:           task.taskId,
+        requiredTaskId,
+        expires:          task.expires,
+      }, true);
+    }));
+
+    // Create TaskDependency entries, each entry implies that taskId is required
+    // by dependentTaskId. This relation is used so taskId can find dependent
+    // tasks when it is resolved.
+    await Promise.all(task.dependencies.map(requiredTaskId => {
+      return this.TaskDependency.create({
+        taskId:           requiredTaskId,
+        dependentTaskId:  task.taskId,
+        expires:          task.expires,
+        relation:         task.dependencyRelation,
+      }, true);
+    }));
+
+    // Load all task dependencies to see if they have been resolved.
+    // We will also check for missing and expiring dependencies.
+    let missing = [];         // Dependencies that doesn't exist
+    let expiring = [];        // Dependencies that expire before deadline
+    let anySatisfied = false; // Track if any dependencies were satisfied
+    await Promise.all(task.dependencies.map(async (requiredTaskId) => {
+      let requiredTask = await this.Task.load({taskId: requiredTaskId}, true);
+
+      // If task is missing, we should report and error
+      if (!requiredTask) {
+        return missing.push(requiredTaskId);
+      }
+
+      // Check if requiredTask expires before the deadline
+      if (task.deadline.getTime() > requiredTask.expires.getTime()) {
```

Comment:
I think it's right... if deadline of the current task is greater than expiration of a dependency something is bad...
So we add the dependency (requiredTask) to `expiring` used for reporting error.